src/View/Embed/EmbedContainer.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace SilverStripe\View\Embed;
+
+use Embed\Extractor;
+use Embed\Embed;
+use Embed\Http\Crawler;
+use Psr\Http\Message\UriInterface;
+use SilverStripe\Core\Manifest\ModuleResourceLoader;
+
+/**
+ * Container for embed/embed v4
+ */
+class EmbedContainer implements Embeddable
+{
+    private ?Embed $embed = null;
+
+    private ?Extractor $extractor = null;
+
+    private ?Crawler $crawler = null;
+
+    private string $url;
+
+    private array $options = [];
+
+    public function __construct(string $url)
+    {
+        $this->url = $url;
+    }
+
+    /**
+     * @return int
+     */
+    public function getWidth()
+    {
+        $code = $this->getExtractor()->code;
+        return $code ? ($code->width ?: 100) : 100;
+    }
+
+    /**
+     * @return int
+     */
+    public function getHeight()
+    {
+        $code = $this->getExtractor()->code;
+        return $code ? ($code->height ?: 100) : 100;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPreviewURL()
+    {
+        $extractor = $this->getExtractor();
+
+        // Use thumbnail url
+        if ($extractor->image) {
+            return (string) $extractor->image;
+        }
+
+        // Default media
+        return ModuleResourceLoader::resourceURL(
+            'silverstripe/asset-admin:client/dist/images/icon_file.png'
+        );
+    }
+
+    /**
+     * @return string
+     */
+    public function getName()
+    {
+        $extractor = $this->getExtractor();
+        if ($extractor->title) {
+            return $extractor->title;
+        }
+        if ($extractor->url instanceof UriInterface) {
+            return basename($extractor->url->getPath());
+        }
+        return '';
+    }
+
+    /**
+     * @return string
+     */
+    public function getType()
+    {
+        $html = $this->getExtractor()->code->html ?? '';
+        if (strpos($html, '<video') !== false) {
+            return 'video';
+        }
+        if (strpos($html, '<iframe') !== false) {
+            return 'rich';
+        }
+        if (strpos($html, '<audio') !== false) {
+            return 'audio';
+        }
+        if (strpos($html, '<img') !== false) {
+            return 'photo';
+        }
+        return 'link';
+    }
+
+    /**
+     * @return bool
+     */
+    public function validate()
+    {
+        return !empty($this->getExtractor()->code->html ?? '');
+    }
+
+    public function setOptions(array $options): Embeddable
```

Comment:
I've updated the interface, get/set options should have been there all along since EmbedResource used them and EmbedShortcodeProvider::handle_shortcode() referenced them

Reply:
Updated